src/linux.cc

```diff
@@ -287,6 +292,57 @@ void update_gateway_info_failure(const char *reason) {
 /* Iface Destination Gateway Flags RefCnt Use Metric Mask MTU Window IRTT */
 #define RT_ENTRY_FORMAT "%63s %lx %lx %x %*d %*d %*d %lx %*d %*d %*d\n"
 
+int update_gateway_info2(void) {
```

Comment:
I see you added `${iface}`. `docker0` stopped showing up too.

Reply:
Pure `${iface}` returns all the network interfaces cards as space and comma separated string.

Reply:
Nope...
```
    gw_iface:[eno1]                                                     
    iface:[eno1]
    iface0:[]
    iface1:[eno1]
    iface2:[]
    iface3:[]
    iface4:[]
    iface5:[]
    iface6:[]
    --------------- 
```
I'm missing my `docker0`. :wink:

Reply:
Because it doesn't have valid destination addrses/mask or gateway.

The following code tests for these destinations and only then puts the NIC.

```c
if (!(dest || mask) && ((flags & RTF_GATEWAY) || !gate)) {
```

Reply:
The following patch will print the docker0

```diff
diff --git a/src/core.cc b/src/core.cc
index 31405de4..ed3ff6b7 100644
--- a/src/core.cc
+++ b/src/core.cc
@@ -1449,9 +1449,8 @@ struct text_object *construct_text_object(char *s, const char *arg, long line,
   END OBJ(gw_iface, &update_gateway_info) obj->callbacks.print =
       &print_gateway_iface;
   obj->callbacks.free = &free_gateway_info;
-  END OBJ(iface, 0) obj->data.s = STRNDUP_ARG;
-  update_gateway_info2(obj->data.s);
-  obj->callbacks.print = &print_gateway_iface;
+  END OBJ(iface, &update_gateway_info2) obj->data.s = STRNDUP_ARG;
+  obj->callbacks.print = &print_gateway_iface2;
   obj->callbacks.free = &gen_free_opaque;
   END OBJ(gw_ip, &update_gateway_info) obj->callbacks.print = &print_gateway_ip;
   obj->callbacks.free = &free_gateway_info;
diff --git a/src/linux.cc b/src/linux.cc
index c636b008..f8798a78 100644
--- a/src/linux.cc
+++ b/src/linux.cc
@@ -292,7 +292,7 @@ void update_gateway_info_failure(const char *reason) {
 /* Iface Destination Gateway Flags RefCnt Use Metric Mask MTU Window IRTT */
 #define RT_ENTRY_FORMAT "%63s %lx %lx %x %*d %*d %*d %lx %*d %*d %*d\n"
 
-int update_gateway_info2(const char *s) {
+int update_gateway_info2(void) {
   FILE *fp;
   char iface[64];
   unsigned long dest;
@@ -322,7 +322,6 @@ int update_gateway_info2(const char *s) {
     if (!(dest || mask) && ((flags & RTF_GATEWAY) || !gate)) {
       snprintf(e_iface, 49, "%s", iface);
     }
-    if (0 == (strcmp(s, ""))) {continue;}
     if (1U == x) {
       snprintf(interfaces_arr[x++], 63, "%s", iface);
       continue;
@@ -397,10 +396,31 @@ int gateway_exists(struct text_object *obj) {
 }
 
 void print_gateway_iface(struct text_object *obj, char *p, unsigned int p_max_size) {
+  (void)obj;
+  snprintf(p, p_max_size, "%s", gw_info.iface);
+
+}
+
+void print_gateway_iface2(struct text_object *obj, char *p, unsigned int p_max_size) {
   long int z = 0;
+  unsigned int x = 1;
+  unsigned int found = 0;
+  char buf[64*64] = {""};
+  char *buf_ptr = buf;
 
-  if (!obj->data.s || (0 == (strcmp(obj->data.s, "")))) {
-    snprintf(p, p_max_size, "%s", gw_info.iface);
+  if (0 == strcmp(obj->data.s, "")) {
+    for (; x < 63; x++) {
+      if (0 == strcmp("", interfaces_arr[x])) {
+        break;
+      }
+      buf_ptr += snprintf(buf_ptr, 63, "%s, ", interfaces_arr[x]);
+      found = 1;
+    }
+    if (1 == found) {
+      --buf_ptr;
+      *(--buf_ptr) = '\0';
+    }
+    snprintf(p, p_max_size, "%s", buf);
     return;
   }
 
diff --git a/src/linux.h b/src/linux.h
index b081ef0b..a50790f4 100644
--- a/src/linux.h
+++ b/src/linux.h
@@ -33,10 +33,11 @@ void print_ioscheduler(struct text_object *, char *, unsigned int);
 void print_laptop_mode(struct text_object *, char *, unsigned int);
 
 int update_gateway_info(void);
-int update_gateway_info2(const char *);
+int update_gateway_info2(void);
 void free_gateway_info(struct text_object *obj);
 int gateway_exists(struct text_object *);
 void print_gateway_iface(struct text_object *, char *, unsigned int);
+void print_gateway_iface2(struct text_object *, char *, unsigned int);
 void print_gateway_ip(struct text_object *, char *, unsigned int);
 
 enum { PB_BATT_STATUS, PB_BATT_PERCENT, PB_BATT_TIME };
```

Reply:
Just replace `update_gateway_info2` with this:

```c
int update_gateway_info2(void) {
  FILE *fp;
  char iface[64];
  unsigned long dest;
  unsigned long gate;
  unsigned long mask;
  unsigned int flags;
  unsigned int x = 1;
  unsigned int z = 1;
  unsigned int skip = 0;

  if ((fp = fopen("/proc/net/route", "r")) == nullptr) {
    update_gateway_info_failure("fopen()");
    return 0;
  }

  /* skip over the table header line, which is always present */
  if (fscanf(fp, "%*[^\n]\n") < 0) {
    fclose(fp);
    return 0;
  }

  while (!feof(fp)) {
    if (fscanf(fp, RT_ENTRY_FORMAT, iface, &dest, &gate, &flags, &mask) != 5) {
      update_gateway_info_failure("fscanf()");
      break;
    }
    if (!(dest || mask) && ((flags & RTF_GATEWAY) || !gate)) {
      snprintf(e_iface, 49, "%s", iface);
    }
    if (1U == x) {
      snprintf(interfaces_arr[x++], 63, "%s", iface);
      continue;
    } else if (0 == strcmp(iface, interfaces_arr[x - 1])) {
      continue;
    }
    for (z = 1; z < 63; z++) {
      if (0 == strcmp(iface, interfaces_arr[z])) {
        skip = 1;
        break;
      }
    }
    if (0 == skip) {
      snprintf(interfaces_arr[x++], 63, "%s", iface);
    }
    skip = 0;
  }
  fclose(fp);
  return 0;
}
```

Reply:
Re-read https://github.com/brndnmtthws/conky/issues/259. This user had a different idea. Suppose we can filter them out later via regexp or conky might have the ability to (via various `if`s).. like `if iface1 != lo`... I don't know... I think we should worry about implementing `${iface}` and `${iface 1}` first... and worry about the rest later... versus hardcoded approach.

I initially thought it was good idea to add options, but now I think we could just use `iface regexp`. Idk. Let me know what you think. Maybe you can support regex here now or do that later?

Reply:
Don't know, I've never written any regex code in my life.